lib/configure-jsx-transform.js

```diff
@@ -0,0 +1,43 @@
+const VersionChecker = require('ember-cli-version-checker');
+
+function requireTransform(transformName) {
+  return require.resolve(transformName);
+}
+
+function hasPlugin(plugins, name) {
+  for (const maybePlugin of plugins) {
+    const plugin = Array.isArray(maybePlugin) ? maybePlugin[0] : maybePlugin;
+    const pluginName = typeof plugin === 'string' ? plugin : plugin.name;
+
+    if (pluginName === name) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+module.exports = function configureJsxTransform(parent) {
+  const options = (parent.options = parent.options || {});
+
+  const checker = new VersionChecker(parent).for('ember-cli-babel', 'npm');
```

Comment:
This is actually utilising ember-cli-babel from the Ember App, which might be problematic if the app is using older Ember that uses babel 5. I would suggest to hook up babel ourselves so that we control how to compile JSX files ourselves.

Reply:
This was what I had in mind - ie use Babel from the app. Are you thinking that we add `ember-cli-babel` as a dependency to `ember-cli-react` and use that?

Reply:
I am actually thinking to use `babel-core` directly instead of relying on `ember-cli-babel`. The advantage of this is that this will work for any app with any version of `ember-cli-babel`. What do you think?

Reply:
hmm... we could add another condition to lib to check if they're using babel5. Is there any difference between babel5 &  babel6? 

It seems to come down to wether or not we're using the app's babel. I would prefer to use app's babel so babel features that are available in React and Ember are consistent. It would be pretty annoying to start using a feature in React then realize that it's not available in the Ember app and vice versa.

Reply:
Babel 6 is incompatible with Babel 5, at least from the configuration. Also Babel 6 removed "require interop" that caused lots of application to have a hard to upgrading last time.

I agree separating Babel config is troublesome but I think there are situations that we want to transpile React files differently with Ember files. Not sure if I am worrying too much :P

@alexgb any input?

Reply:
@alexgb what we're trying to figure out right now is what configuration API does ember-cli-react need to provide to the users. 

@pswai suggested that we should have a separate babel compiler and configure that compiler. 

I suggested that we should just add `jsx` transform to app's babel compiler. This way, whatever custom transforms people add, they'll be available in both JSX and regular JS.

What do you think?

Reply:
> If I understand correctly, you're saying that when using broccoli-react the babel configuration options applied to the host app have no affect on JSX files?

Yes

> It also exposes configuration for its babel transformation that we should be able to use to add additional plugins here. Is that a viable path forward here?

This is where this PR started. If you look at the start of this PR, the first commit is https://github.com/AltSchool/ember-cli-react/pull/42/commits/da25380d50776827b3003c0fd6f0d413f0b3725f which makes it possible to configure broccoli-react.

> But I do see how it's important to have rest spread and class properties support in JSX files, so let's figure out the right solution.

I suggested to remove `broccoli-react` in favour of using using app's babel because on you start using these features in React, you'll probably want to use them in the Ember app. Then you find it doesn't work in the Ember app because your configuration is only applied to `ember-cli-react`. Then you end up with something like this,

```js
const EmberApp = require('ember-cli/lib/broccoli/ember-app');

const plugins = [
  'transform-decorators-legacy',
  'transform-object-rest-spread',
  'transform-class-properties'
];

module.exports = function(defaults) {
  let app = new EmberApp(defaults, {
    babel: {
      plugins
    },
    'ember-cli-react': {
      babelOptions: {
        plugins
      },
    },
  });

  return app.toTree();
};
```

If we're ok with this, then I can just revert the changes and that I made in last few commits and push the original PR.

Reply:
Gotcha. Thanks for walking me through your thinking. So the main decision is between using the host app's babel to process JSX or an isolated `babel-core`, which feels like a choice between a two evils:

  1. If we use the host app's babel, then `ember-cli-react` has a somewhat brittle binding to host app's babel version to add JSX support, which we must maintain as babel is updated.
  2. If we use isolated `babel-core` then we force the user to know how to configure both babel6, as used by `broccoli-react`, and their own babel version.

I think I'm coming around to using the host app's babel, but don't feel strongly. If we go this route I think it would be wise for us to add a peer dependency on `ember-cli-babel` so that we don't imply support for future unreleased versions of `babel`. Also, how is this currently working on babel 5? The tests seem to be passing on versions of `ember-cli` older than 2.12, which should be using babel 5, but your code only seems to inject JSX support on babel 6 or babel 7.

Reply:
> If we go this route I think it would be wise for us to add a peer dependency on ember-cli-babel so that we don't imply support for future unreleased versions of babel.

👍 

> Also, how is this currently working on babel 5? The tests seem to be passing on versions of ember-cli older than 2.12, which should be using babel 5, but your code only seems to inject JSX support on babel 6 or babel 7.

Are you sure it's using babel 5? The fact that it doesn't change babel dependency suggests to me that it's using whatever is installed before `ember-try` runs. I could be wrong though. We might be able to add a test for babel 5.

Reply:
Yep, I think you're right, which concerns me because we do test and indicate support for Ember versions down to 1.13 and it's unfortunate that ember-try just exercises ember and ember-data versions, but not the full set of ember-cli dependencies for a particular release.

So yes, would be good to test babel5 compatibility, but not sure how to test that end to end.